weierstrass/src/projective.rs

```diff
@@ -0,0 +1,695 @@
+        let yy_m_bzz3 = yy - &bz3_part; // 16
+        let yy_p_bzz3 = yy + &bz3_part; // 17
+
+        let z3 = self.z.double() + &self.z; // 19, 20
+        let bxz_part = (C::EQUATION_B * &xz_pairs) - &(z3 + &xx); // 18, 21, 22
+        let bxz3_part = bxz_part.double() + &bxz_part; // 23, 24
+        let xx3_m_zz3 = xx.double() + &xx - &z3; // 25, 26, 27
+
+        let mut ret = Self {
+            x: (yy_p_bzz3 * &xy_pairs) - &(yz_pairs * &bxz3_part), // 28, 32, 33
+            y: (yy_p_bzz3 * &yy_m_bzz3) + &(xx3_m_zz3 * &bxz3_part), // 29, 30, 31
+            z: (yy_m_bzz3 * &yz_pairs) + &(xy_pairs * &xx3_m_zz3), // 34, 35, 36
+        };
+        ret.conditional_assign(self, other.is_identity());
+        ret
+    }
+
+    /// Doubles this point.
+    pub fn double(&self) -> Self {
+        // We implement the exception-free point doubling formula from
+        // Renes-Costello-Batina 2015 (Algorithm 6). The comments after each line
+        // indicate which algorithm steps are being performed.
+
+        let xx = self.x.square(); // 1
+        let yy = self.y.square(); // 2
+        let zz = self.z.square(); // 3
+        let xy2 = (self.x * &self.y).double(); // 4, 5
+        let xz2 = (self.x * &self.z).double(); // 6, 7
+
+        let bzz_part = (C::EQUATION_B * &zz) - &xz2; // 8, 9
+        let bzz3_part = bzz_part.double() + &bzz_part; // 10, 11
+        let yy_m_bzz3 = yy - &bzz3_part; // 12
+        let yy_p_bzz3 = yy + &bzz3_part; // 13
+        let y_frag = yy_p_bzz3 * &yy_m_bzz3; // 14
+        let x_frag = yy_m_bzz3 * &xy2; // 15
+
+        let zz3 = zz.double() + &zz; // 16, 17
+        let bxz2_part = (C::EQUATION_B * &xz2) - &(zz3 + &xx); // 18, 19, 20
+        let bxz6_part = bxz2_part.double() + &bxz2_part; // 21, 22
+        let xx3_m_zz3 = xx.double() + &xx - &zz3; // 23, 24, 25
+
+        let y = y_frag + &(xx3_m_zz3 * &bxz6_part); // 26, 27
+        let yz2 = (self.y * &self.z).double(); // 28, 29
+        let x = x_frag - &(bxz6_part * &yz2); // 30, 31
+        let z = (yz2 * &yy).double().double(); // 32, 33, 34
+
+        Self { x, y, z }
+    }
+
+    /// Returns `self - other`.
+    pub fn sub(&self, other: &Self) -> Self {
+        self.add(&other.neg())
+    }
+
+    /// Returns `self - other`.
+    fn sub_mixed(&self, other: &AffinePoint<C>) -> Self {
+        self.add_mixed(&other.neg())
+    }
+
+    /// Returns `[k] self`.
+    fn mul(&self, k: &Scalar<C>) -> Self {
+        let k: C::UInt = (*k).into();
+        let k = k.to_le_byte_array();
+
+        let mut pc = [Self::default(); 16];
+        pc[0] = Self::IDENTITY;
+        pc[1] = *self;
+
+        for i in 2..16 {
+            pc[i] = if i % 2 == 0 {
+                pc[i / 2].double()
+            } else {
+                pc[i - 1].add(self)
+            };
+        }
+
+        let mut q = Self::IDENTITY;
+        let mut pos = C::UInt::BIT_SIZE - 4;
+
+        loop {
+            let slot = (k[(pos >> 3) as usize] >> (pos & 7)) & 0xf;
```

Comment:
This logic is slightly different between `p256` and `p384`, and I'm not quite clear how to make it generic.

The version above is from `p384`. The `p256` version looks like this:

```rust
let slot = (k[31 - (pos >> 3) as usize] >> (pos & 7)) & 0xf;
```

Reply:
Aah, it was the endianness of the serialized scalar. I've went with little endian, which makes things slightly simpler.